test/e2e/storage/drivers/csi.go

```diff
@@ -222,6 +221,9 @@ func (h *hostpathCSIDriver) PrepareTest(f *framework.Framework) (*testsuites.Per
 
 		ginkgo.By(fmt.Sprintf("deleting the driver namespace: %s", ns2))
 		tryFunc(deleteNamespaceFunc(f.ClientSet, ns2, framework.DefaultNamespaceDeletionTimeout))
+		// we need to remove the cleanupHandle at last so as if AfterEach and AfterSuite races with
+		// each other and AfterEach(or defer) run gets killed then it still cleanups.
+		framework.RemoveCleanupAction(h.cleanupHandle)
```

Comment:
The original logic was "if cleanupFunc fails, then don't try to run it again because it'll just fail again". Is this logic flawed?

How can cleanupFunc be invoked and get interrupted while it is still running? That part is not clear to me; mention a race doesn't explain it.

Reply:
Here is how Ginkgo handles interrupts:

1. It starts running `AfterSuite` hooks in a separate goroutine (this includes cleanupAction hooks)
2. Once `AfterSuite` hook is done executing it calls `os.Exit(1)`  on test suite.

So how cleanupFunc() that runs via `defer` in test can be interrupted is:
1. `cleanupFunc` starts running via `defer` (or `AfterEach` hook) but first thing that function does is to remove cleanupHandle from `framework.RemoveCleanupAction`.
2. Test suite receives interrupt from user and `AfterSuite` block starts executing 
3. remember that while `cleanupFunc` is running in goroutine#1, `AfterSuite` is running concurrently in goroutine#2.  
4. AfterSuite hook has bunch of `CleanupActions` it needs to run which were registered via `framework.AddCleanupAction(cleanupFunc)` but once `cleanupFunc` starts executing via `defer` in the test, it will remove the cleanupHandle from framework's aftersuite hooks.
5. So if `AfterSuite` did not had anything to run (because those actions were removed via `framework.RemoveCleanupAction` then it will simply go to the last `framework.AfterEach` action and call `os.Exit(1)`
6. So if `os.Exit(1)` is called before `cleanupFunc` has a chance to finish in `defer`, it will not complete.

Naturally to see this race,you will have to interrupt the test suite but even then it may not happen always.  But I will bring this in another PR. This change does not belong here.

Reply:
That's a useful explanation, please include it in the commit message so that it gets recorded for others.

Reply:
I will add this in the follow up PR. cheers.